### PR TITLE
Videoplayer code cleanup & race condition fix

### PR DESCRIPTION
**Pull request by MaxKellermann**

### CODE REVIEW

File: xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitleStream.cpp [hunk 1]
```diff
@@ -73,7 +72,7 @@ bool CDVDSubtitleStream::Open(const std::string& strFile)
         totalread += read;
     } while (read > 0);
 
-    delete pInputStream;
+    pInputStream.reset();
```
Comment (garbear) [inline xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitleStream.cpp]: returning should automatically reset, correct? this line can be dropped

File: xbmc/cores/VideoPlayer/DVDDemuxSPU.cpp [hunk 2]
```diff
@@ -365,7 +365,7 @@ CDVDOverlaySpu* CDVDDemuxSPU::ParseRLE(CDVDOverlaySpu* pSPU, uint8_t* pUnparsedD
   unsigned int i_x, i_y;
 
   // allocate a buffer for the result
-  unsigned __int16* p_dest = (unsigned __int16*)pSPU->result;
+  uint16_t* p_dest = (uint16_t*)pSPU->result;
```
Comment (garbear) [inline xbmc/cores/VideoPlayer/DVDDemuxSPU.cpp]: mind changing to reinterpret_cast while you're at it?

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp [hunk 3]
```diff
@@ -890,16 +890,17 @@ void CVideoPlayer::OpenDefaultStreams(bool reset)
...
 
   // open video stream
-  streams = m_SelectionStreams.Get(STREAM_VIDEO, PredicateVideoPriority);
   valid   = false;
-  for(SelectionStreams::iterator it = streams.begin(); it != streams.end() && !valid; ++it)
+
+  for (const auto &stream : m_SelectionStreams.Get(STREAM_VIDEO, PredicateVideoPriority))
   {
-    if(OpenStream(m_CurrentVideo, it->demuxerId, it->id, it->source, reset))
+    if(OpenStream(m_CurrentVideo, stream.demuxerId, stream.id, stream.source, reset)) {
```
Comment (garbear) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: bracket on second line to match our code style

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp [hunk 4]
```diff
@@ -90,16 +90,14 @@ using namespace KODI::MESSAGING;
...
-
-    m_Streams.erase(m_Streams.begin() + i);
-  }
+  auto new_end =
+    std::remove_if(m_Streams.begin(),
+                   m_Streams.end(),
+                   [type, source](const SelectionStream &stream){
+                     return (type == STREAM_NONE || stream.type == type) &&
+                       (source == 0 || stream.source == source);
+                   });
```
Comment (garbear) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: This can stay as is, but I think that readability is improved if newlines are single-indented by structuring these awkward lambas like this:

```c++
auto new_end = std::remove_if(m_Streams.begin(), m_Streams.end(),
  [type, source](const SelectionStream &stream)
  {
    return (type == STREAM_NONE || stream.type == type) &&
           (source == 0 || stream.source == source);
  });
```
other examples from a PR of mine: https://github.com/xbmc/xbmc/blob/master/xbmc/input/joysticks/generic/ButtonMapping.cpp#L141

**Comment by garbear** [comment]
Thanks for the cleanup. Code improvements are nice, but IMO what's more important is that it deepens your understanding of what the hell's going on in some of Kodi's larger and more complex areas :) Keep up the good work.

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp (hunk 4 above)
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: this entire class needs a re-write

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp [hunk 5]
```diff
@@ -117,6 +117,7 @@ SelectionStream& CSelectionStreams::Get(StreamType type, int index)
 
 std::vector<SelectionStream> CSelectionStreams::Get(StreamType type)
 {
+  CSingleLock lock(m_section);
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: the actual problem is this this is not supposed to be called from other threads

**Comment by FernetMenta** [comment]
please submit in smaller junks. VideoPlayer is in the process of transformation. There are lots of areas under construction and doing cosmetics to this code is just a change without a step forward. 
There are many changes I agree with but with others I don't.

File: xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitleStream.cpp (hunk 1 above)
Comment (MaxKellermann) [inline xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitleStream.cpp]: On 2016/03/24 21:33, Garrett Brown <notifications@github.com> wrote:
> > @@ -73,7 +72,7 @@ bool CDVDSubtitleStream::Open(const std::string& strFile)
> >          totalread += read;
> >      } while (read > 0);
> >  
> > -    delete pInputStream;
> > +    pInputStream.reset();
> 
> returning should automatically reset, correct? this line can be dropped

Yes, but the old code did it here, so I did the same, to retain old
behavior.  Sometimes, there can be an advantage in freeing an object
earlier in the scope, and in this case, I wasn't sure if that early
delete was intentional.  Of course, I'll drop it if you prefer so.

File: xbmc/cores/VideoPlayer/DVDDemuxSPU.cpp (hunk 2 above)
Comment (MaxKellermann) [inline xbmc/cores/VideoPlayer/DVDDemuxSPU.cpp]: On 2016/03/24 21:34, Garrett Brown <notifications@github.com> wrote:
> > @@ -365,7 +365,7 @@ CDVDOverlaySpu* CDVDDemuxSPU::ParseRLE(CDVDOverlaySpu* pSPU, uint8_t* pUnparsedD
> >    unsigned int i_x, i_y;
> >  
> >    // allocate a buffer for the result
> > -  unsigned __int16* p_dest = (unsigned __int16*)pSPU->result;
> > +  uint16_t* p_dest = (uint16_t*)pSPU->result;
> 
> mind changing to reinterpret_cast while you're at it?

Yes, but IMHO such things have to be done in a different commit.  It
would be confusing to do two things at once, it would be more
difficult to review.

**Comment by MaxKellermann** [comment]
On 2016/03/24 22:43, Rainer Hochecker <notifications@github.com> wrote:
> please submit in smaller junks. VideoPlayer is in the process of transformation. There are lots of areas under construction and doing cosmetics to this code is just a change without a step forward. 
> There are many changes I agree with but with others I don't.

Splitting to smaller junks is difficult without knowing which ones you
agree with.  Can you just cherry-pick the ones you agree with, and
then we'll talk about the rest?

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp (hunk 5 above)
Comment (MaxKellermann) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: What does that mean? What is "other thread"? Does that mean the class doesn't need protection at all, and the mutex should be removed? There is protection currently, and it's incorrect. Either remove it completely, or fix it (like I did).

**Comment by FernetMenta** [comment]
ok

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp (hunk 5 above)
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: VideoPlayer has its own thread and only this thread is supposed to access this method. It was designed only for use in OpenDefaultStreams: https://github.com/xbmc/xbmc/pull/700/commits/240692b2a3a4b9675bb4c972e745d7a1e8e1c3de

Unfortunately folks started to use it for different purpose like exposing information to the GUI. When calling this method the vector is copied twice which is a performance issue.

Comment (MaxKellermann) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: On 2016/03/25 08:26, Rainer Hochecker <notifications@github.com> wrote:
> VideoPlayer has its own thread and only this thread is supposed to access this method. It was designed only for use in OpenDefaultStreams: https://github.com/xbmc/xbmc/pull/700/commits/240692b2a3a4b9675bb4c972e745d7a1e8e1c3de

I see.  So that's the current state, which we must deal with.  Which
means my commit is necessary.

If you don't want my commit, plan B is to remove all uses of
CSelectionStreams from other parts of Kodi, then document
CSelectionStreams to be **not** thread-safe, and remove locking
overhead.

What do you want me to do?

> Unfortunately folks started to use it for different purpose like exposing information to the GUI. When calling this method the vector is copied twice which is a performance issue.

Twice?  It is copied once (i.e. data from "m_Streams" is copied to
local variable "streams").  The "return streams" is not a copy,
it's "moved" to the caller.  That is just a pointer exchange.  Some
C++ compilers might even be able to eliminate this move (maybe
gcc+clang do this already, I'm not sure).

Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: > If you don't want my commit, plan B is to remove all uses of

Nope. Only this particular method does not require locking. See your commit message. It's only VideoPlayer thread that can change the vector. The short term solution is not calling this method from outside.

**Comment by MaxKellermann** [comment]
Rebased with the following changes:
- indentation of lambdas changed as suggested
- explicit release() call removed, even though I think it was better before

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp (hunk 5 above)
Comment (MaxKellermann) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: So instead of the race condition fix, will you accept a patch that adds API documentation stating that using this method makes an instances thread-unsafe?

Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: Usage of this method from any other thread needs to be removed. See how it is used. The streams are copied to the local variable and then a second time (subs in this case).

```
bool CVideoPlayer::GetStreamDetails(CStreamDetails &details)
{
  if (m_pDemuxer)
  {
    std::vector<SelectionStream> subs = m_SelectionStreams.Get(STREAM_SUBTITLE);
```

This snippet shows the next no-go. The outside world must not access demuxer

Comment (MaxKellermann) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: I understand, the only thing I don't understand is what you want me to do. I made three suggestions, and you didn't agree to any of these. I'd like to post something that meets your agreement, but it appears impossible.

Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: >  what you want me to do

remove the calls to m_SelectionStreams.Get in methods called from the outside.

Comment (MaxKellermann) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: Please tell me where class CSelectionStreams is accessed from outside of the player thread. I found nothing. Or more precise: it is used in the CVideoPlayer constructor, at a time where the player thread has not yet been spawned. That is safe.

Am I right to assume that locking in CSelectionStreams is completely unnecessary?

**Comment by MaxKellermann** [comment]
Rebased and dropped the patch "VideoPlayer: fix race condition in Get(StreamType)", because @FernetMenta asked me to rewrite it. I believe that all of the locking was never necessary (see https://github.com/xbmc/xbmc/pull/9442#discussion_r57465958), and thus a reconsideration of the class is necessary.

As soon as we have a result, I'll implement the solution. Until then, please merge the remaining commits.

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp (hunk 5 above)
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: > Please tell me where class CSelectionStreams is accessed from outside of the player thread

see the snippet in my post above: i.e.
CVideoPlayer::GetStreamDetails or
CVideoPlayer::GetAudioStreamInfo
(there is more)

> Am I right to assume that locking in CSelectionStreams is completely unnecessary?

If a user wants to change audio or subtitle track the UI needs to have access to the currently available streams. The UI also accesses this object for getting info to display on the OSD.
Having the GUI poll player data frequently is not ideal. It is much better to provide an extra object for the UI and player updates this object after something has changed.
That's the vision and we don't need to get to final in a single iteration.

Comment (MaxKellermann) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: On 2016/03/25 19:08, Rainer Hochecker <notifications@github.com> wrote:
> see the snippet in my post above: i.e.
> CVideoPlayer::GetStreamDetails or
> CVideoPlayer::GetAudioStreamInfo
> (there is more)

OK, so these methods are called from outside the player thread.  I
thought what you meant was users of CSelectionStreams from classes
other than VideoPlayer.  I see.

I will check these methods and see what I can do to implement lockless
read-only access.  (Out of this PR's scope.)

**Comment by MaxKellermann** [comment]
Rebase & changed indent according to https://github.com/xbmc/xbmc/pull/9442#discussion-diff-57383380

**Comment by wsnipex** [comment]
jenkins build this please

**Comment by FernetMenta** [comment]
we agreed that I cherry-pick from this PR. no need to put load on Jenkins here.

**Comment by MaxKellermann** [comment]
Rebased, dropping the patches that were merged by cherry-picking PR https://github.com/xbmc/xbmc/pull/9454

Darwin/Android builds probably still fail due to their broken system headers. To be checked.

**Comment by MaxKellermann** [comment]
So, this turns out not to be broken system headers, but a kludge in utils/Regexp.h which looks so ugly, it's surprising that it breaks only Android and Darwin. The offending code (abbreviated):

    namespace PCRE {
    #include <pcre.h>
    }

This tries to be clever by moving all `pcre.h` definitions into a separate namespace. The problem is that `pcre.h` includes `stdlib.h`, thus all if `stdlib.h`'s definitions are now also living in the `PCRE` namespace. This obviously breaks anything that searches for `stdlib.h` definitions in the root namespace.

Ouch!

So this code has always been there, added by the extraordinarily huge (and undocumented) commit https://github.com/xbmc/xbmc/commit/45285e8

**Comment by MartijnKaijser** [comment]
@MaxKellermann yes that commit is some breakpoint in time :) Should you wish to dig deeper the continued history is here https://github.com/xbmc/xbmc-antiquated

**Comment by garbear** [comment]
I've run into this problem before. I solved it by creating a wrapper header, e.g. `pcre_wrapper.h`, that included pcre.h's headers outside the namespace, then included pcre inside the namespace. e.g.

```c++
// Include all headers used by pcre.h
#include <stdlib.h>
namespace PCRE
{
#include <pcre.h>
}
```

**Comment by MaxKellermann** [comment]
@garbear, had the same idea to work around this for now. Though I'd like to get rid of this fragile namespace hack. What purpose does it serve?

**Comment by garbear** [comment]
It's plausible that pcre.h contains one or more symbols that conflict with Kodi in the global namespace

**Comment by MaxKellermann** [comment]
I submitted https://github.com/xbmc/xbmc/pull/9494 to fast-track the workaround and the include cleanup. I think the rest was not-yet-accepted by @FernetMenta, so after merging 9494, we should discuss the rest of this PR.

**Comment by MaxKellermann** [comment]
After #9494 has been merged, I've rebased this branch. Please discuss. (@FernetMenta?)

**Comment by FernetMenta** [comment]
@MaxKellermann I'll look at it tomorrow. Need to fix some bugs first.

**Comment by MaxKellermann** [comment]
After merging #9530, this is a rebase with 4 remaining unmerged commits.

**Comment by FernetMenta** [comment]
Adding overrides is correct but this interface needs cleanup first. Means looking at the methods one by one and decide if those still make sense. In order to do so history and git blame are valuable tools. https://github.com/xbmc/xbmc/pull/9442/commits/9b1032926ff799aa826cfd88bdb0e0a9d93c12af would make this work harder. This is why I don't pick it in this stage.

ee89f85c434f8f230d392860f32d96ac79d2ff49 why don't you make the class final? That would make it obvious that it is not supposed to inherit from.

I am ok with the other two in selection streams if @ace20022 is as well.

**Comment by MaxKellermann** [comment]
Having the "overrides" makes it easier to change the API, because you'll get compiler errors if you remove a `virtual` from the base class without removing them from all derived classes.

I'll add `final` soon. I wasn't sure if MSVC knows `final` - I've become cautious with C++11 keywords because your MSVC version is too old for some C++11 features.

**Comment by FernetMenta** [comment]
I won't take the overrides commit for the reasons given. Feel free to review individual methods and submit PRs, but I won't accept this bulk operation.

**Comment by afedchin** [comment]
@MaxKellermann don't worry about vs, it knows about final

**Comment by MaxKellermann** [comment]
Dropped the `override` commit from this PR, and added `final` commit.

**Comment by FernetMenta** [comment]
thanks
jenkins build this please

**Comment by ace20022** [comment]
> I am ok with the other two in selection streams if @ace20022 is as well.

 I have no objections.

**Comment by FernetMenta** [comment]
Windows fails with:

```
7>..\..\xbmc\cores\VideoPlayer\VideoPlayer.cpp(120): error C2039: 'back_inserter': is not a member of 'std' [c:\jenkins\slave\workspace\WIN-32\project\VS2010Express\XBMC.vcxproj]
```

**Comment by ace20022** [comment]
> Windows fails with:

```#include <iterator>``` is missing

**Comment by MaxKellermann** [comment]
Oops. Should be fixed now. Please try again.

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by FernetMenta** [comment]
thanks again!
